relay-server/src/actors/processor.rs

```diff
@@ -2358,14 +2364,68 @@ impl EnvelopeProcessorService {
     }
 
     /// Computes the sampling decision on the incoming transaction.
-    fn compute_sampling_decision(&self, state: &mut ProcessEnvelopeState) {
-        state.sampling_result = utils::get_sampling_result(
-            self.inner.config.processing_enabled(),
-            Some(&state.project_state),
-            state.sampling_project_state.as_deref(),
-            state.envelope().dsc(),
-            state.event.value(),
-        );
+    fn compute_sampling_decision(
+        processing_enabled: bool,
+        sampling_config: Option<&SamplingConfig>,
+        event: Option<&Event>,
+        root_sampling_config: Option<&SamplingConfig>,
+        dsc: Option<&DynamicSamplingContext>,
+    ) -> SamplingResult {
+        if (sampling_config.is_none() || event.is_none())
+            && (root_sampling_config.is_none() || dsc.is_none())
+        {
+            return SamplingResult::NoMatch;
+        }
```

Comment:
Early return.

I don't like that the type system doesn't represent how `sampling_config` and `event` is related. and how `root_sampling_config` and `dsc` is connected.

Perhaps some new struct which encapsulates both the configuration and rules, and the instance to match on. Would be a separate PR of course.